pep-0497-externals.txt

```diff
@@ -0,0 +1,460 @@
+inference layer being implemented or specified. Alternatively, a separate PEP
+could be developed in tandem with this proposal to hold the apppropriate level
+of detail.
+
+
+Types of environment and vendor
+-------------------------------
+
+Anaconda is not an operating system. Neither is yum, neither is apt. It is the
+maintainer of the packaging system which is pertinent, and the current operating
+system is an imperfect indicator of the relevant package manager.
+
+
+Directories, categories and locations of dependency files
+---------------------------------------------------------
+
+The proposed mechanism allows introspection of the standard location for
+packages. On *nix environments, these would be (for example), /usr/bin,
+/usr/local/include etc (see https://www.gnu.org/prep/standards/html_node/Directory-Variables.html
+for a discussion of standard directories). Non-standard dependency locations
+would be supported only via the underlying platform's mechanisms for exposing
+these details, such as PATH, LD_LIBRARY_PATH etc. This PEP does not include
+generalised configuration and specification of dependencies in non-standard
+or customised locations.
+
+The Python installation tools should honour those standard mechanisms.
+The details are per-platform and not specified by this specification
+(see Platform Abstraction for more on this point).
+
+Virtual Environment Concerns
+----------------------------
+
+This should work fine with virtual environments. Since the PEP draws the line
+at performing external package installation, it is up to package management
+tools to decide how to separate concerns between systemwide installation
+and virtual environment installation. For example, a system such as
+Anaconda which is heavily based around virtual environments is well capable
+of handling the installation of external dependencies into a virtual environment
+context.
+
+Alternative Syntax Concepts
+==========================
+
+Concept 0: Separated external and non-external dependencies
+-----------------------------------------------------------
+
+This is an early-stage syntax. Concept 1 replaces it as the current
+"front runner" as implementation candidate and is included in the abstract.
+
+The advantage of Concept 0 is that is clearly separates the non-Python
+dependency specification into an isolated part of the configuration arguments,
+reducing the potential for confusion between external and internal dependencies.
+This may assist during implementation as no 'sniffer' will be needed to
+determine whether a named package is internal or external.
+
+::
+
+    setup(
+        name='uses_scipy_with_atlas_extension'
+        install_requires='yaml, scipy[atlas]',
+        setup_requires='pyfortran',
+        external={
+          setup=[
+            'bin!fortran:dist_name=='ubuntu',
+            'include!kernel.h',
+            'include!liblas.h',
+          ],
+          install=[
+              'lib!liblas1.so'
```

Comment:
We don't want to add the dependency automatically because the developer
might want to find a way to remove it, statically link it, or figure out a
suitable distro independent way of describing it, rather than including a
distro specific dependency in the wheel metadata.

It definitely makes sense for the build system to detect and report
discrepancies between stated intent and actual build results when it comes
to external dependencies, but they should still be made explicit rather
than left implicit.